### PR TITLE
Documentation/erc721 contracts

### PR DESCRIPTION
**Pull request by GuillaumeCz**
Improvements / completions on the ERC721 contracts. 

To do: 
- [x] IERC721Receiver
- [x] ERC721Holder

### CODE REVIEW

**Comment by GuillaumeCz** [comment]
I have some questions in order to complete this PR:
- Is there a specific way to document `internal` and `private` functions ? 
- Is documentation is required for `constructors` ? 

Thanks in advance ;)

**Comment by frangio** [comment]
Thank you @GuillaumeCz! I love this work.

> Is there a specific way to document internal and private functions ?

Private functions are not displayed in the docs site, and are not available for users, so we don't have any requirements to document them. It would be fine if you left them untouched in this PR.

I don't think we have any specific way to document internal functions. The site will show an "internal" label, so you can assume the user knows they're internal.

image


> Is documentation is required for constructors ?

Yes, absolutely. Even if only a short trivial sentence, we would like to have something so that it doesn't look like this in the site:

image

**Comment by GuillaumeCz** [comment]
I'm not very sure of the `onERC721Received` functions documentation... 
Please let me know if there's a better way to do it ! ;)

File: contracts/token/ERC721/IERC721Receiver.sol [hunk 1]
```diff
@@ -6,19 +6,19 @@ pragma solidity ^0.6.0;
...
+     *
+     * The selector to be returned can be obtained as `this.onERC721Received.selector`.
+     *
+     * This function may throw to revert and reject the transfer.
+     *
+     * The ERC721 contract address is always the message sender.
+     *
+     * Requirements:
+     *
+     * - This function must return the function selector, otherwise the caller will revert the transaction.
```
Comment (nventuro) [inline contracts/token/ERC721/IERC721Receiver.sol:21]:
```suggestion
     * @dev Whenever an {IERC721} `tokenId` token is transferred to this contract via {IERC721-safeTransferFrom}
     * by `operator` from `from`, this function is called.
     *
     * It must return its Solidity selector to confirm the token transfer. If any other value is returned or
     * the interface is not implemented by the recipient, the transfer will be reverted.
     *
     * The selector can be obtained in Solidity with `IERC721Receiver.onERC721Received.selector`;
```

File: contracts/token/ERC721/ERC721Holder.sol [hunk 2]
```diff
@@ -3,6 +3,20 @@ pragma solidity ^0.6.0;
...
+     * The selector to be returned can be obtained as `this.onERC721Received.selector`.
+     *
+     * This function may throw to revert and reject the transfer.
+     *
+     * The ERC721 contract address is always the message sender.
+     *
+     * Requirements:
+     *
+     * - This function must return the function selector, otherwise the caller will revert the transaction.
+     */
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721Holder.sol:19]:
```suggestion

    /**
     * @dev See {IERC721Receiver-onERC721Received}.
     * 
     * Always returns `IERC721Receiver.onERC721Received.selector`.
     */
```

File: contracts/token/ERC721/ERC721Holder.sol [hunk 3]
```diff
@@ -3,6 +3,20 @@ pragma solidity ^0.6.0;
 import "./IERC721Receiver.sol";
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721Holder.sol:4]:
```suggestion
 /*
  * @dev Implementation of the {IERC721Receiver} interface.
  *
  * Accepts all token transfers: make sure the contract is able to use its tokens with {IERC721-safeTransferFrom}, {IERC721-approve} or {IERC721-setApprovalForAll}.
  */
```

File: contracts/token/ERC721/ERC721.sol [hunk 4]
```diff
@@ -359,40 +289,46 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
...
-     * which is called upon a safe transfer, and return the magic value
-     * `bytes4(keccak256("onERC721Received(address,address,uint256,bytes)"))`; otherwise,
-     * the transfer is reverted.
-     * @param to The address that will own the minted token
-     * @param tokenId uint256 ID of the token to be minted
+     * @dev Safely mints `tokenId` and transfers it to `to`.
+     *
+     * Requirements:
+     *
+     * - `tokenId` should not exist.
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol:296]:
```suggestion
     * - `to` cannot be zero.
     * - `tokenId` must not exist.
```

File: contracts/token/ERC721/ERC721.sol [hunk 5]
```diff
@@ -359,40 +289,46 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
...
-     * @param _data bytes data to send along with a safe transfer check
+     * @dev Safely mints `tokenId` and transfers it to `to`.
+     *
+     * Requirements:
+     *
+     * - `tokenId` should not exist.
+     * - `to` cannot be zero.
+     * - If `to` refers to a smart contract, it must implement {IERC721Receiver-onERC721Received}, which is called upon a safe transfer.
+     *
+     * Emits a {Transfer} event.
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol:314]:
```suggestion
     * @dev Same as {_safeMint-address-uint256-}, with an additional `data` parameter which is forwarded
     * in {IERC721Receiver-onERC721Received} to contract recipients.
```

File: contracts/token/ERC721/ERC721.sol [hunk 6]
```diff
@@ -433,11 +374,14 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
...
-     * @dev Internal function to transfer ownership of a given token ID to another address.
-     * As opposed to {transferFrom}, this imposes no restrictions on msg.sender.
-     * @param from current owner of the token
-     * @param to address to receive the ownership of the given token ID
-     * @param tokenId uint256 ID of the token to be transferred
+     * @dev Transfers `tokenId` from `from` to `to`.
+     *  As opposed to {transferFrom}, this imposes no restrictions on msg.sender.
+     *
+     * Requirements:
+     * - `to` cannot be zero.
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol:381]:
```suggestion
     *
     * - `to` cannot be zero.
```

File: contracts/token/ERC721/ERC721.sol [hunk 7]
```diff
@@ -288,69 +231,56 @@ contract ERC721 is Context, ERC165, IERC721, IERC721Metadata, IERC721Enumerable
...
-     * @param tokenId uint256 ID of the token to be transferred
-     * @param _data bytes data to send along with a safe transfer check
+     * @dev Safely transfers `tokenId` token from `from` to `to`, checking first that contract recipients
+     * are aware of the ERC721 protocol to prevent tokens from being forever locked.
+     *
+     * `_data` is additional data, it has no specified format and it is sent in call to `to`.
+     *
+     * Requirements:
+     *
+     * - `from`, `to` cannot be zero.
```
Comment (nventuro) [inline contracts/token/ERC721/ERC721.sol:256]:
```suggestion
     * - `from` cannot be the zero address.
     * - `to` cannot be the zero address.
```

We should make this change in all relevant functions

File: contracts/token/ERC721/ERC721.sol (hunk 4 above)
Comment (nventuro) [reply contracts/token/ERC721/ERC721.sol:296]: We should change the 'should' phrasing to 'must' in all relevant functions

**Comment by GuillaumeCz** [comment]
Thank you very much @nventuro for your review ! I was quite struggling with internal functions documentation ;) thanks for your help and suggestions ! 

Sorry for being long for the PR, I was quite busy at the end of the lockdown ;)
I hope that the provided fixes will do the job !

**Pull request merged**